lib/Service/OpenAiAPIService.php

```diff
@@ -85,6 +87,114 @@ public function getPromptHistory(string $userId, int $type): array {
 		return $this->promptMapper->getPromptsOfUser($userId, $type);
 	}
 
+	/**
+	 * Clear prompt history for a prompt type
+	 * @param string $userId
+	 * @param int $type
+	 */
+	public function clearPromptHistory(string $userId, int $type): void {
+		$this->promptMapper->deleteUserPromptsByType($userId, $type);
+	}
+
+	/**
+	 * @param string $userId
+	 */
+	private function hasOwnOpenAiApiKey(string $userId): bool {
+		if(!$this->isUsingOpenAi()){
+			return false;
+		}
+
+		if($this->config->getUserValue($userId, Application::APP_ID, 'api_key') !== ''){
+			return true;
+		}
+
+		return false;
+	}
+
+
+	/**
+	 * Check whether quota is exceeded for a user
+	 * @param string|null $userId
+	 * @param int $type
+	 * @return bool
+	 * @throws Exception
+	 */
+	public function isQuotaExceeded(?string $userId, int $type): bool {
+		if($userId === null){
+			return false;
+		}
+		
+		if(!array_key_exists($type, Application::QUOTA_TYPES)){
+			throw new Exception('Invalid quota type');
+		}
+
+		if($this->hasOwnOpenAiApiKey($userId)){
+			// User has specified own OpenAI API key, no quota limit:
+			return false;
+		}
+
+		// Get quota limits
+		$quota = intval(json_decode($this->config->getAppValue(Application::APP_ID, 'quotas', json_encode(Application::DEFAULT_QUOTAS)))[$type]->value);
+
+		if($quota === 0){
+			//  Unlimited quota:
+			return false;
+		}
+
+		$quotaPeriod = intval($this->config->getAppValue(Application::APP_ID, 'quota_period', Application::DEFAULT_QUOTA_PERIOD));
+		
+		try {
+			$quotaUsage= $this->quotaUsageMapper->getQuotaUnitsOfUserInTimePeriod($userId, $type, $quotaPeriod);
+		} catch (DoesNotExistException | MultipleObjectsReturnedException | Exception | RuntimeException $e) {
+			throw new Exception('Could not retrieve quota usage for user: ' . $userId . ' and quota type: ' . $type . '. Error: ' . $e->getMessage());
+		}
+		
+		return $quotaUsage >= $quota;
+	}
+
+	/**
+	 * @param string $userId
+	 * @return array
+	 */
+	public function getUserQuotaInfo(string $userId): array {
+		// Get quota limits (if the user has specified an own OpenAI API key, no quota limit, just supply default values as fillers)
+		$quotas = $this->hasOwnOpenAiApiKey($userId) ?
+				Application::DEFAULT_QUOTAS :
+				json_decode($this->config->getAppValue(Application::APP_ID, 'quotas', json_encode(Application::DEFAULT_QUOTAS)));
```

Comment:
Also, note that json_decode uses StdClass by default instead of an associative array so $quotas will now have the type `\StdClass[]|array[]` which is kind of unwieldy

Reply:
This is a good reason why not using an AI coding assistant and looking up the function's docs instead is often a good idea ;)

Reply:
Huh, that's an interesting default behavior, btw. To keep things in line with js syntax?

Reply:
Thanks! Fixed in https://github.com/nextcloud/integration_openai/pull/54/commits/44720e0aebfd3f549a6d9bd48a78df2872f26110